dbt-adapters/src/dbt/include/global_project/macros/materializations/models/incremental/merge.sql

```diff
@@ -61,36 +61,23 @@
     {%- set dest_cols_csv = get_quoted_csv(dest_columns | map(attribute="name")) -%}
 
     {% if unique_key %}
-        {% if unique_key is sequence and unique_key is not string %}
-            delete from {{ target }}
-            using {{ source }}
-            where (
-                {% for key in unique_key %}
-		    {% set source_unique_key = (source ~ "." ~ key) | trim %}
-		    {% set target_unique_key = (target ~ "." ~ key) | trim %}
-                    {{ equals(source_unique_key, target_unique_key) }}
-                    {{ "and " if not loop.last}}
-                {% endfor %}
-                {% if incremental_predicates %}
-                    {% for predicate in incremental_predicates %}
-                        and {{ predicate }}
-                    {% endfor %}
-                {% endif %}
-            );
-        {% else %}
-            delete from {{ target }}
-            where (
-                {{ unique_key }}) in (
-                select ({{ unique_key }})
-                from {{ source }}
-            )
-            {%- if incremental_predicates %}
-                {% for predicate in incremental_predicates %}
-                    and {{ predicate }}
-                {% endfor %}
-            {%- endif -%};
-
+        {% if unique_key is string %}
+        {% set unique_key = [unique_key] %}
         {% endif %}
+
+        {%- set unique_key_str = unique_key|join(', ') -%}
+
+        delete from {{ target }}
+        where ({{ unique_key_str }}) in (
+            select distinct {{ unique_key_str }}
```

Comment:
I think the `distinct` here is unnecessary and would make it take longer since it needs to reduce dupes. I don't have proof of that, but I generally don't use `distinct` when using a `where <x> in <y>` clause.

Reply:
Hey Mike, so this is actually turned out to be a really interesting problem to me. `distinct` should be faster in theory, but in practice, cardinality of the subquery's result set is key.

Higher cardinality means the DBMS will do far more scans of the target result set than needed.

Lower cardinality means the deduplication step is irrelevant and hence a performance lowering event.

We can't predict the cardinality for the unique key strings that users pick, but we do note in docs that users should employ `delete+insert` for more efficient merges in the context of high cardinality sets. Hence, let's default to `distinct` as always present for this workflow and avoid the possibility that indexing/algorithms under the hood fail to prune extraneous unmatched rows in the result set.